text/0000-is-sorted.md

```diff
@@ -0,0 +1,272 @@
+duplicate code. Optimally, we would have a trait that represents sequential
+data structures and would only add `is_sorted` and friends to said trait. We
+don't have such a trait as of now; so `Iterator` is the next best thing. Slices
+deserve special treatment due to the reasons mentioned above (popularity and
+`sort()`).
+
+
+### `Iterator::while_sorted`, `is_sorted_until`, `sorted_prefix`, `num_sorted`, ...
+[In the issue on the main repository](https://github.com/rust-lang/rust/issues/44370),
+concerns about completely consuming the iterator were raised. Some alternatives,
+such as [`while_sorted`](https://github.com/rust-lang/rust/issues/44370#issuecomment-327873139),
+were suggested. However, consuming the iterator is neither uncommon nor a
+problem. Methods like `count()`, `max()` and many more consume the iterator,
+too. [One comment](https://github.com/rust-lang/rust/issues/44370#issuecomment-344516366) mentions:
+
+> I am a bit skeptical of the equivalent on Iterator just because the return
+> value does not seem actionable -- you aren't going to "sort" the iterator
+> after you find out it is not already sorted. What are some use cases for this
+> in real code that does not involve iterating over a slice?
+
+As mentioned above, `Iterator` is the next best thing to a trait representing
+sequential data structures. So to check if a `LinkedList`, `VecDeque` or
+another sequential data structure is sorted, one would simply call
+`collection.iter().is_sorted()`. It's likely that this is the main usage for
+`Iterator`'s `is_sorted` methods. Additionally, code like
+`if v.is_sorted() { v.sort(); }` is not very useful:  `sort()` already runs in
+O(n) for already sorted arrays.
+
+Suggestions like `is_sorted_until` are not really useful either: one can easily
+get a subslice or a part of an iterator (via `.take()`) and call `is_sorted()`
+on that part.
+
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+
+### Is `Iterator::is_sorted_by_key` useless?
+
+[One comment in the corresponding issue](https://github.com/rust-lang/rust/issues/44370#issuecomment-327740685)
+mentions that `Iterator::is_sorted_by_key` is not really necessary, given that
+you can simply call `map()` beforehand. This is true, but it might still be
+favourable to include said function for consistency and ease of use. The
+standard library already hosts a number of sorting-related functions all of
+which come in three flavours: *raw*, `_by` and `_by_key`. By now, programmers would
+probably expect there to be an `is_sorted_by_key` as well.
+
+
+### Add `std::cmp::is_sorted` instead
+
+As suggested [here](https://github.com/rust-lang/rust/issues/44370#issuecomment-345495831),
+one could also add this free function (plus the `_by` and `_by_key` versions)
+to `std::cmp`:
+
+```rust
+fn is_sorted<C>(collection: C) -> bool
+where
+    C: IntoIterator,
+    C::Item: Ord,
+```
+
+This can be seen as a better design as it avoids the question about which data
```

Comment:
Perhaps if Rust has a built in operator for function composition and made point-free notation easy, but that is not the case, so let's keep the RFC as is.